server/sonar-server/src/main/java/org/sonar/server/rule/RuleUpdater.java

```diff
@@ -264,34 +273,85 @@ private void deleteOrUpdateParameters(DbSession dbSession, RuleUpdate update, Ru
 
       if (value != null) {
         // Update linked active rule params or create new one
-        for (ActiveRuleDto activeRuleDto : activeRules.get(customRule)) {
-          for (ActiveRuleParamDto activeRuleParamDto : activeRuleParams.get(activeRuleDto)) {
-            if (activeRuleParamDto.getKey().equals(key)) {
-              dbClient.activeRuleDao().updateParam(dbSession, activeRuleDto, activeRuleParamDto.setValue(value));
-            } else {
-              dbClient.activeRuleDao().insertParam(dbSession, activeRuleDto, ActiveRuleParamDto.createFor(ruleParamDto).setValue(value));
-            }
-          }
-        }
+        from(activeRuleParamsByActiveRule.keySet())
+          .filter(new UpdateOrInsertActiveRuleParams(dbSession, dbClient, ruleParamDto, activeRuleParamsByActiveRule))
+          .toList();
       } else {
         // Delete linked active rule params
-        for (ActiveRuleDto activeRuleDto : activeRules.get(customRule)) {
-          for (ActiveRuleParamDto activeRuleParamDto : activeRuleParams.get(activeRuleDto)) {
-            if (activeRuleParamDto.getKey().equals(key)) {
-              dbClient.activeRuleDao().deleteParam(dbSession, activeRuleDto, activeRuleParamDto);
-            }
-          }
-        }
+        from(activeRuleParamsByActiveRule.values())
+          .filter(new DeleteActiveRuleParams(dbSession, dbClient, key))
+          .toList();
       }
       paramKeys.add(key);
     }
   }
 
+  private static class ActiveRuleParamToActiveRule implements Function<ActiveRuleParamDto, ActiveRuleDto> {
+    private final Map<Integer, ActiveRuleDto> activeRuleById;
+
+    private ActiveRuleParamToActiveRule(Map<Integer, ActiveRuleDto> activeRuleById) {
+      this.activeRuleById = activeRuleById;
+    }
+
+    @Override
+    public ActiveRuleDto apply(@Nonnull ActiveRuleParamDto input) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this "Nonnull" annotation to honor the overridden method's contract. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)